src/metrics/include/metrics/operation.hpp

```diff
@@ -0,0 +1,260 @@
+
+    OperationContextT(OperationContextT<ClockSource>&& other) noexcept
+        : _op{std::move(other._op)},
+          _started{std::move(other._started)},
+          _event{std::move(other._event)},
+          _isClosed{std::exchange(other._isClosed, true)} {}
+
+    ~OperationContextT() {
+        if (!_isClosed) {
+            BOOST_LOG_TRIVIAL(warning)
+                << "Metrics not reported because operation '" << _op->getOpName()
+                << "' being run by actor '" << _op->getActorName()
+                << "' did not close with success() or failure().";
+        }
+    }
+
+    /**
+     * Increments the counter for the number of iterations. This method only needs to be called if
+     * an actor is periodically reporting its operations. By default an `iters = 1` value is
+     * automatically reported.
+     */
+    void addIterations(count_type iters) {
+        _event.iters += iters;
+    }
+
+    /**
+     * Increments the counter for the number of documents inserted, modified, deleted, etc.
+     */
+    void addDocuments(count_type ops) {
+        _event.ops += ops;
+    }
+
+    /**
+     * Increments the size in bytes of the number of documents inserted, etc.
+     */
+    void addBytes(count_type size) {
+        _event.size += size;
+    }
+
+    /**
+     * Increments the counter for the number of write errors, transient transaction errors, etc.
+     */
+    void addErrors(count_type errors) {
+        _event.errors += errors;
+    }
+
+    /**
+     * Report the operation as having succeeded.
+     *
+     * @warning After calling success() it is illegal to call any methods on this instance.
+     */
+    void success() {
+        reportOutcome(OutcomeType::kSuccess);
+    }
+
+    /**
+     * Report the operation as having failed. An `errors > 0` value won't be reported unless
+     * addErrors() has already been called.
+     *
+     * @warning After calling failure() it is illegal to call any methods on this instance.
+     */
+    void failure() {
+        reportOutcome(OutcomeType::kFailure);
+    }
+
+    /**
+     * Don't report the operation.
+     *
+     * @warning After calling discard() it is illegal to call any methods on this instance.
+     */
+    void discard() {
+        _isClosed = true;
+    }
+
+private:
+    void reportOutcome(OutcomeType outcome) {
+        auto finished = ClockSource::now();
+        _event.duration = finished - _started;
+        _event.outcome = outcome;
+
+        if (_event.iters == 0) {
+            // We default the event to represent a single iteration of a loop if addIterations() was
+            // never called.
+            _event.iters = 1;
+        }
+
+        _op->reportAt(finished, std::move(_event));
+        _isClosed = true;
+    }
+
+    v1::OperationImpl<ClockSource>* const _op;
+    const time_point _started;
+
+    OperationEvent<ClockSource> _event;
+    bool _isClosed = false;
+};
+
+
+template <typename ClockSource>
+class OperationT {
```

Comment:
we haven't been super consistent with it, but I think make all these classes `final` - they are not designed to be subclassed and it seems like good hygiene to document that intention.

Reply:
Done.